File: src/frontend/src/api/tables.js
import api from './index'

/**
 * 数据表查看 API。
 *
 * 后端契约：
 *   GET /api/v1/tables/{table_name}?page=1&size=20&keyword=
 *   返回 { total: number, rows: [...] }
 *
 * 由后端白名单表配置执行筛选、排序和分页，避免在浏览器下载全量数据。
 */
export const TABLES = [
  {
    name: 'raw_observations',
    label: '原始观测',
    desc: '当前可信范围内已采集、可追溯的原始来源记录',
    requiresScope: true,
    columns: [
      { prop: 'source_name', label: '来源', sortable: true, minWidth: 170 },
      { prop: 'record_type', label: '记录类型', sortable: true, minWidth: 140 },
      { prop: 'observed_date', label: '观测日期', sortable: true, width: 120 },
      { prop: 'source_record_key', label: '来源记录键', sortable: true, minWidth: 220 },
      { prop: 'entity_mapping_status', label: '实体映射', sortable: true, width: 120 },
      { prop: 'category_mapping_status', label: '品类映射', sortable: true, width: 120 },
      { prop: 'quality_status', label: '质量状态', sortable: true, width: 120 },
      { prop: 'source_url', label: '来源链接', sortable: true, minWidth: 220 },
    ],
  },
  {
    name: 'metric_observations',
    label: '快照指标',
    desc: '当前可信范围下、绑定快照与口径版本的指标事实',
    requiresScope: true,
    columns: [
      { prop: 'snapshot_id', label: '快照 ID', sortable: true, minWidth: 210 },
      { prop: 'metric_key', label: '指标', sortable: true, minWidth: 190 },
      { prop: 'entity_type', label: '实体层级', sortable: true, width: 110 },
      { prop: 'entity_key', label: '实体键', sortable: true, minWidth: 180 },
      { prop: 'value', label: '值', sortable: true, width: 130 },
      { prop: 'unit', label: '单位', sortable: true, width: 90 },
      { prop: 'quality_status', label: '质量状态', sortable: true, width: 120 },
      { prop: 'metric_version', label: '口径版本', sortable: true, minWidth: 130 },
      { prop: 'calculated_at', label: '计算时间', sortable: true, minWidth: 180 },
    ],
  },
  {
    name: 'store_operations',
    label: '门店经营数据',
    desc: 'Excel 导入的真实销售、成本与合同经营记录',
    columns: [
      { prop: 'record_date', label: '日期', sortable: true, width: 120 },
      { prop: 'brand_id', label: '品牌 ID', sortable: true, width: 120 },
      { prop: 'store_id', label: '门店 ID', sortable: true, width: 170 },
      { prop: 'sales_amount', label: '销售额', sortable: true, width: 120 },
      { prop: 'order_count', label: '订单数', sortable: true, width: 100 },
      { prop: 'customer_price', label: '客单价', sortable: true, width: 100 },
      { prop: 'customer_flow', label: '客流量', sortable: true, width: 100 },
      { prop: 'rent', label: '租金', sortable: true, width: 100 },
      { prop: 'rent_to_sales_ratio', label: '租售比(%)', sortable: true, width: 120 },
      { prop: 'sales_per_sqm', label: '坪效', sortable: true, width: 100 },
      { prop: 'contract_end', label: '合同到期', sortable: true, width: 120 },
      { prop: 'data_source', label: '数据来源', sortable: false, minWidth: 180 },
    ],
  },
]

export function getTableConfig(name) {
  return TABLES.find((t) => t.name === name) || TABLES[0]
}

/**
 * 查询数据表。
 * @returns {Promise<{ total: number, rows: object[] }>}
 */
export async function fetchTableData(tableName, { page = 1, size = 20, keyword = '', sortProp = '', sortOrder = '', scopeId = '' } = {}) {
  const { data } = await api.get(`/v1/tables/${tableName}`, {
    params: { page, size, keyword: keyword.trim(), sort_prop: sortProp, sort_order: sortOrder, scope_id: scopeId },
  })
  return data
}

File: src/backend/brandpulse/api/tables.py
"""受限的数据表浏览 API：白名单字段、服务端筛选、排序与分页。"""
from dataclasses import dataclass
from typing import Dict, Tuple

from fastapi import APIRouter, HTTPException, Query
from sqlalchemy import text

from brandpulse.api.dashboard import to_jsonable
from brandpulse.db_clients.postgres_client import PostgresClient

router = APIRouter(prefix="/api/v1/tables", tags=["tables"])


@dataclass(frozen=True)
class TableSpec:
    fields: Tuple[str, ...]
    searchable_fields: Tuple[str, ...]
    default_order: str
    requires_scope: bool = False


TABLE_SPECS: Dict[str, TableSpec] = {
    "raw_observations": TableSpec(
        fields=("source_name", "record_type", "observed_date", "source_record_key", "entity_mapping_status", "category_mapping_status", "quality_status", "source_url"),
        searchable_fields=("source_name", "record_type", "source_record_key", "source_url"),
        default_order="observed_date DESC, source_name ASC, source_record_key ASC",
        requires_scope=True,
    ),
    "metric_observations": TableSpec(
        fields=("snapshot_id", "metric_key", "entity_type", "entity_key", "value", "unit", "quality_status", "metric_version", "calculated_at"),
        searchable_fields=("snapshot_id", "metric_key", "entity_type", "entity_key", "quality_status"),
        default_order="calculated_at DESC, metric_key ASC",
        requires_scope=True,
    ),
    "store_operations": TableSpec(
        fields=("op_id", "brand_id", "store_id", "record_date", "sales_amount", "order_count", "customer_price", "customer_flow", "rent", "store_area", "rent_to_sales_ratio", "sales_per_sqm", "contract_end", "data_source"),
        searchable_fields=("brand_id", "store_id", "data_source"),
        default_order="record_date DESC, store_id ASC",
    ),
}


@router.get("/{table_name}")
def browse_table(
    table_name: str,
    page: int = Query(default=1, ge=1),
    size: int = Query(default=20, ge=1, le=100),
    keyword: str = Query(default="", max_length=128),
    sort_prop: str = Query(default="", max_length=64),
    sort_order: str = Query(default="", pattern="^(|ascending|descending)$"),
    scope_id: str = Query(default="", max_length=64),
):
    spec = TABLE_SPECS.get(table_name)
    if not spec:
        raise HTTPException(status_code=404, detail="不支持的数据表")

    conditions = []
    params = {"limit": size, "offset": (page - 1) * size}
    if spec.requires_scope:
        if not scope_id.strip():
            raise HTTPException(status_code=422, detail="该数据表必须选择可信监测范围")
        conditions.append("scope_id = :scope_id")
        params["scope_id"] = scope_id.strip()
    if keyword.strip():
        conditions.append("(" + " OR ".join(f"CAST({field} AS TEXT) ILIKE :keyword" for field in spec.searchable_fields) + ")")
        params["keyword"] = f"%{keyword.strip()}%"
    where = " WHERE " + " AND ".join(conditions) if conditions else ""

    if sort_prop in spec.fields and sort_order:
        order = f"{sort_prop} {'ASC' if sort_order == 'ascending' else 'DESC'} NULLS LAST"
    else:
        order = spec.default_order
    fields = ", ".join(spec.fields)
    client = PostgresClient()
    with client.engine.connect() as conn:
        total = conn.execute(text(f"SELECT COUNT(*) FROM {table_name}{where}"), params).scalar_one()
        rows = conn.execute(text(f"""
            SELECT {fields} FROM {table_name}{where}
            ORDER BY {order} LIMIT :limit OFFSET :offset
        """), params).mappings().all()
    return {"total": total, "rows": to_jsonable([dict(row) for row in rows])}
